### ISSUE TITLE
reset-yaw goofs things up

### ISSUE THREAD

**Issue by rpavlik** [OSVR/OSVR-Core#75]
Basically, you can reproduce by doing reset yaw particularly with a large angle - you'll get at least some degree of axis swapping.

I think it's because we're just applying a rotation instead of doing a "basis change".

cc @russell-taylor

**Comment by russell-taylor** [comment]
I've put it on my ToDo list after getting VRPN transferred and getting a video-based LED tracker.  Let me know if it should be higher priority.

**Comment by russell-taylor** [comment]
In the transformation hierarchy I am used to thinking about, the current behavior is like adjusting the sensor-to-hmd transformation and the desired behavior is like adjusting the room-to-world transformation.  These occur at opposite ends of the transformation hierarchy, so I expect moving the rotation (appropriately transformed into the correct space) to the other end of the transformation hierarchy to do the trick.

Essentially, we want this rotation to happen in a more-inner nesting than the tracker -- as a final rotation around the world-space Y coordinate.  This is essentially "flying" per the Holloway/Robinett paper -- rotating the room with respect to the world.

Checking this: Changing the outer -X rotation from 180 degrees to 145; the resulting rotation around the HMD up axis rotates around the world-space up axis rather than the local green up axis.  Rotation around the nose rotates around an axis in the world-space plane rather than the local Z axis.

Because we're constructing the transformations from the point of view of the eye and heading towards the world, we end up needing one more transformation past the tracker to do this yaw adjustment -- to adjust around the world-space Y axis.  The rest of the transforms that we have are done on our side of the tracker.

*** What I'd like to do is make another child UNDER the tracker object, but I don't see how to do that.  If I name the tracker something other than "child", it does not seem to behave as a tracker anymore.  If I name both the tracker and the other node "child", the second will overwrite the first because of how Json works.  I think if we named the tracker object "tracker" rather than "child", then we could express the desired feature.  I don't know if the code in OSVR would deal with this expression if we could make it.

I think you may be faster than me at making OSVR internals allow different names and have an additional transform past the current one, but I'm willing to give it a try if you'd like.  Just let me know.

**Comment by rpavlik** [comment]
So we do actually have the ability to apply "postrotate" etc - more than just what I have shown in config files. hasn't really been documented yet, but the code that parses it is here:
https://github.com/OSVR/OSVR-Core/blob/master/src/osvr/Common/JSONTransformVisitor.cpp#L117

**Comment by russell-taylor** [comment]
Adding a "postrotate" key as a direct parent of the tracker child seems to do the trick -- it maintains the correct local orientation variation but let you set an offset in the tracker's local coordinates.  I love the "null fix" where you've already done something we needed!

Still to do is figuring out the appropriate transformation to apply to get the desired effect.  Will need to transfer the desired rotation through the intervening transformations and move where this gets inserted into the Json.

**Comment by russell-taylor** [comment]
Example section from the file, for documenting purposes.  I've left-shifted the inserted parts to highlight them:

      {
         "destination" : "/me/head",
         "source" : {
            "calibration" : true,
            "child" : {
               "child" : {
                  "rotate" : {
                     "axis" : "-x",
                     "degrees" : 180
                  },
                  "child" : {
                     "changeBasis" : {
                        "x" : "x",
                        "y" : "-z",
                        "z" : "-y"
                     },
"child" : {
   "postrotate" : {
     "axis" : "z",
     "radians" : 0.5
   },
                     "child" : {
                        "sensor" : 0,
                        "tracker" : "/org_opengoggles_bundled_Multiserver/OSVRHackerDevKit0"
                     }
}
                  }
               }
            }

         }
      }

**Comment by russell-taylor** [comment]
Okay, so it also works if you do postrotate and wrap the entire object, which means that just changing the following does the trick:

    -            newLayer["rotate"]["radians"] = -yaw;
    -            newLayer["rotate"]["axis"] = "y";
    +            newLayer["postrotate"]["radians"] = -yaw;
    +            newLayer["postrotate"]["axis"] = "y";

**Comment by russell-taylor** [comment]
A fix for this has been pushed to the "improved-reset-yaw" branch.  However, the app does not compile in that branch -- it and diagnose-rotation both are filed under experimental and can't find getRoutes() anymore.  I'll issue a pull request for that branch and leave this ticket open so that we remember to finish porting those apps to the 0.2 code.  Once the port is done, the yaw axis-scrambling should be gone.

I did test this fix in a pre-0.2 branch and they fixed things.

**Comment by rpavlik** [comment]
This was fixed long ago, thanks!
